### PR TITLE
New package: openloops

### PR DESCRIPTION
**Pull request by iarspider**

### CODE REVIEW

**Comment by iarspider** [comment]
Is the failure in the tests known or something is wrong with my code? I have tested it on Centos7 + gcc 8 and it worked fine...

File: var/spack/repos/builtin/packages/openloops/package.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            copy(join_path(os.path.dirname(__file__), 'sft2.coll'), 'lcg.coll')
+
+    def build(self, spec, prefix):
+        scons = Executable('./scons')
+        scons('generator=1', 'compile=2')
+
+    def build_processes(self, spec, prefix):
+        ol = Executable('./openloops')
+        processes = self.spec.variants['processes'].value
+        if self.spec.variants['compile_extra']:
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/openloops/package.py]:
```suggestion
        if 'compile_extra' in self.spec:
```

File: var/spack/repos/builtin/packages/openloops/package.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                print('gfortran_f_flags = -ffree-line-length-none ' +
+                      '-fdollar-ok -mcmodel=medium', file=f)
+
+        if self.spec.satisfies('@:1.999.999 processes=lcg.coll'):
+            copy(join_path(os.path.dirname(__file__), 'sft1.coll'), 'lcg.coll')
+        elif self.spec.satisfies('@2:2.999.999 processes=lcg.coll'):
+            copy(join_path(os.path.dirname(__file__), 'sft2.coll'), 'lcg.coll')
+
+    def build(self, spec, prefix):
+        scons = Executable('./scons')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/openloops/package.py:107]: Interesting, scons is distributed with the source code?

File: var/spack/repos/builtin/packages/openloops/package.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        for k in env.keys():
+            if k.startswith('SPACK_'):
+                spack_env.append(k)
+
+        spack_env = ' '.join(spack_env)
+        is_intel = self.spec.satisfies('%intel')
+        njobs = self.spec.variants['num_jobs'].value
+
+        with open('openloops.cfg', 'w') as f:
+            print('[OpenLoops]', file=f)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/openloops/package.py]:
```suggestion
            f.write('[OpenLoops]\n')
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by vvolkl** [comment]
This recipe works well for me, thanks @iarspider. As Adam suggested, the `print('[OpenLoops]', file=f)` syntax won't work with older python versions so that will be good to change.

File: var/spack/repos/builtin/packages/openloops/package.py (hunk 2 above)
Comment (iarspider) [reply var/spack/repos/builtin/packages/openloops/package.py:107]: Yes

File: var/spack/repos/builtin/packages/openloops/package.py (hunk 3 above)
Comment (iarspider) [reply var/spack/repos/builtin/packages/openloops/package.py]: I was assuming that Spack targets python3 (as far as I can see, when possible it chooses python3), and "behind the scenes" does `from __future__ import ...`.

**Comment by iarspider** [comment]
The list of available processes is different in OpenLoops 1.x and 2.x. Should I handle this somehow?

File: var/spack/repos/builtin/packages/openloops/package.py (hunk 3 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/openloops/package.py]: Nope. We equally target Python 2.6+ and 3.5+ with no magic future imports. Also, https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files still recommends using `f.write` even in Python 3.8.

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
> The list of available processes is different in OpenLoops 1.x and 2.x. Should I handle this somehow?

I can't think of a way to properly handle this off the top of my head. Maybe @alalazo would have an idea. I don't think a multi-valued variant's values can be specified by version, just like variants can't be specified by version.

**Comment by alalazo** [comment]
@adamjstewart Until we get variants that might depend on other parameters (like versions) I think the way to deal with that is conflicts. It will be tedious though given the quantity of processes in this package.

**Comment by iarspider** [comment]
@adamjstewart @alalazo can we just leave this (validation of process list) to the users? We have a reasonable default (`lhc.coll`, maintained by OpenLoops developers), and the installation will fail if a requested process is missing.

**Pull request merged**